pandas/tests/groupby/transform/test_transform.py

```diff
@@ -149,7 +149,7 @@ def test_transform_broadcast(tsframe, ts):
             assert_fp_equal(res[col], agged[col])
 
     # group columns
-    grouped = tsframe.groupby({"A": 0, "B": 0, "C": 1, "D": 1}, axis=1)
+    grouped = tsframe.groupby({"A": 0, "B": 0, "C": 1, "D": 1}, axis=1, dropna=False)
```

Comment:
a few of these tests assume that groupby doesn't drop `NaN`s which was ok before b/c `dropna` didn't propagate

`dropna` propagates now and defaults to`True` so needs to be set

Reply:
The keys here don't contain `NaN` though, right? Why is dropna impacting this?

Reply:
Fair comment - it's a bug introduced in this patch. Reverted change to the test in the recent commit since as you say it should work w/o `dropna` arg. 

The changes I made break this branch of `DataFrameGroupBy._transform_general`, specifically in the `else` (the error is reshape mismatch)
https://github.com/pandas-dev/pandas/blob/a09259b5699b0f2667761356cb3b22ef43ef37f0/pandas/core/groupby/generic.py#L1331-L1349
I haven't got a handle on how to rewrite the conditions here - would welcome any pointers!